submissions/ustymchyk/js-post-oop/index.js

```diff
@@ -0,0 +1,212 @@
+        } else {
+            console.log('Sorry, we are traditional');
+        }
+    }
+}
+
+class Pet extends Creature {
+    #happySound;
+    #angrySound;
+    #energy;
+
+    constructor(name, gender, age, happySound, angrySound) {
+        if (new.target === Pet) {
+            throw new TypeError("Cannot construct Abstract instances directly");
+        }
+
+        super(name, gender, age);
+
+        this.#happySound = happySound;
+        this.#angrySound = angrySound;
+        this.#energy = 5;
+    }
+
+    play() {
+        if (this.#energy) {
+            this.#energy--;
+            this._talk(this.#happySound);
+        } else {
+            this._talk(this.#angrySound);
+            this._rest();
+        }
+    }
+
+    pat() {
+        this._talk(this.#happySound);
+    }
+
+    feed() {
+        this.#energy++;
+        this._talk(this.#happySound);
+    }
+
+    introduce() {
+        return `${this.#happySound} ${this.name} ${this.#happySound} ${this.#happySound} ${this.gender} ${this.#happySound} ${this.age}`;
+    }
+
+    _talk(sound) {
+        console.log(sound);
+    }
+
+    _rest() {
+        this.#energy = 5;
+    }
+}
+
+// I use _ instead of # in methods for prototype inheritance.
+class Cat extends Pet {
+    constructor(name, gender, age) {
+        super(name, gender, age, 'meow', 'shhhh');
+        this.species = 'cat';
+    }
+}
+
+class Dog extends Pet {
+    constructor(name, gender, age) {
+        super(name, gender, age, 'wof', 'grrrr');
+        this.species = 'dog';
+    }
+}
+
+class Human extends Creature {
+    #friends = [];
+    #sympathy = [];
+    #contacts = [];
+
+    constructor(name, gender, age) {
+        super(name, gender, age);
+        this.species = 'human';
+        this.spouse = null;
+    }
+
+    talkWith(otherHuman) {
+        if (otherHuman instanceof Human) {
+            if (!this.#contacts.includes(otherHuman.name)) {
+                this._createContact(otherHuman);
+                otherHuman.talkWith(this);
+            }
+        } else {
+            console.log('I can speak only with human');
+        }
+    }
+
+    makeProposal(otherHuman) {
+        if (!this.spouse && otherHuman.answerProposal(this)) {
+            this.spouse = otherHuman.name;
+        }
+    }
+
+    answerProposal(otherHuman) {
+        if (!this.spouse && this.#sympathy.includes(otherHuman.name)) {
+            this.spouse = otherHuman.name;
+
+            return true;
+        }
+    }
+
+    getFriendsNames() {
+        return this.#friends.join('; ');
+    }
+
+    introduce() {
+        const intro = `Hello! My name is ${this.name}, I'm ${this.age}. `;
+        let friends = '';
+        let spouse = '';
+
+        if (this.#friends.length) {
+            friends = `There is my friends: ${this.getFriendsNames()}. `;
+        }
+
+        if (this.spouse) {
+            spouse = `And i have a spouse ${this.spouse}; `;
+        }
+
+        return intro + friends + spouse;
+
+    }
```

Comment:
It is not illegal that child classes use properties introduced by a base class.
However for the sake of method overloading demonstration let's make all inhabitants intorducing themselves in a similar manner. In terms of hierarchy and OOP principles let's implement it so that every class manages only the part with properties introduced by such base or child class.